### DISCUSSION TITLE
toSession doesn't work after upgarde from v0.10.0 to v.0.13.0

### DISCUSSION THREAD

**Discussion by amirbrzz1** [ory/kratos#3458] [Q&A] [answered]
Hi gays.

I recently upgraded Kratos from v0.10.0 to v0.13.0, and also upgraded @ory/client to the v1.1.50
due to this upgrade, toSession doesn't work anymore, and returns : 
**### The request could not be authorized reason:No valid session credentials found in the request** 

It is strange here when I switch back to Kratos v0.10.0, toSession works correctly

 NPM packages version:
        "@ory/cli": "^0.2.2",
        "@ory/client": "^1.1.50",
        "@ory/kratos-client": "^0.13.1"

```
async login(identity: Models.Api.Ply_AccountRequest.AuthenticateRequest): Promise<Ory.Models.LoginSuccessResult | undefined> {
        return await sdk.createBrowserLoginFlow({ aal: 'aal1'}, { headers: { 'Accept': 'application/json' } })
            .then(async result => {
                const csrf_token = extractCsrfToken(result);
                const cookie = result.headers['set-cookie'];
                var payload: UpdateLoginFlowWithPasswordMethod = {
                    method: "password",
                    password_identifier: identity.username,
                    identifier: identity.username,
                    csrf_token: csrf_token,
                    password: identity.password,
                  }
                
                return await sdk
                    .updateLoginFlow(
                        { flow: flowId, updateLoginFlowBody: payload, cookie: cookie},
                        { headers: { withCredentials: true, "Accept": "application/json", "Content-Type": "application/json" } })
                    .then(async response => {

                        await sdk.toSession({ cookie: cookie },{
                            headers: {
                                'Content-Type': 'application/json',
                                'X-Session-Cookie': `${cookie}`
                            }
                        }
                        ).catch(err => {
                            **//The request could not be authorized reason:No valid session credentials found in the request.**
                            throw new HttpException('', 100);
                        }).then(result => {
                        });
                    })
                    .catch(err => {
                        throw new InvalidCredentialsException(err, message);
                    })
            });

    }
```

kratos.yaml

```
version: v0.13.0

dsn: memory

serve:
  public:
    base_url: http://127.0.0.1:4433/
    cors:
      enabled: true
      allowed_origins:
        - https://localhost:10000
        - https://localhost:10002 
        - https://localhost:10001
      allowed_methods:
        - POST
        - GET
        - PUT
        - PATCH
        - DELETE
      allowed_headers:
        - Authorization
        - Content-Type
        - Cookie
      exposed_headers:
        - Content-Type
        - Set-Cookie
      allow_credentials: true
      debug: true
  admin:
    base_url: http://kratos:4434/

selfservice:
  default_browser_return_url: http://127.0.0.1:4455/
  allowed_return_urls:
    - http://127.0.0.1:4455

  methods:
    password:
      enabled: true
    totp:
      config:
        issuer: Kratos
      enabled: true
    lookup_secret:
      enabled: true
    link:
      enabled: true
    code:
      enabled: true

  flows:
    error:
      ui_url: http://127.0.0.1:4455/error

    settings:
      ui_url: http://127.0.0.1:4455/settings
      privileged_session_max_age: 15m
      required_aal: highest_available

    recovery:
      enabled: true
      ui_url: http://127.0.0.1:4455/recovery
      use: code

    verification:
      enabled: true
      ui_url: http://127.0.0.1:4455/verification
      use: code
      after:
        default_browser_return_url: http://127.0.0.1:4455/

    logout:
      after:
        default_browser_return_url: http://localhost:3000/sign-in

    login:
      ui_url: http://localhost:3000/sign-in
      lifespan: 10m

    registration:
      lifespan: 10m
      ui_url: http://localhost:3000/registration
      after:
        password:
          hooks:
            - hook: session
            - hook: show_verification_ui

log:
  level: debug
  format: text
  leak_sensitive_values: true

secrets:
  cookie:
    - PLEASE-CHANGE-ME-I-AM-VERY-INSECURE
  cipher:
    - 32-LONG-SECRET-NOT-SECURE-AT-ALL

ciphers:
  algorithm: xchacha20-poly1305

hashers:
  algorithm: bcrypt
  bcrypt:
    cost: 8

identity:
  default_schema_id: default
  schemas:
    - id: default
      url: file:///etc/config/kratos/identity.schema.json

courier:
  smtp:
    connection_uri: smtps://test:test@mailslurper:1025/?skip_ssl_verify=true

session:  
  # sessions can't be refreshed until 8 hours before they expire.
  earliest_possible_extend : 8h
  cookie:    
    same_site: Lax


```

these are my docker containers created by quickstart page.
![image](https://github.com/ory/kratos/assets/71702933/c4da2805-3d84-475a-bcc0-9f5a65d4d3b7)

my we app runs on :
**https://localhost:10000**
I tried it by **http://localhost:10000** too

and Ory kratos on :
**http://localhost:4433**

**Comment by vinckr** [comment]
Hello @amirbrzz1 
Please make sure you are running the latest version of the SDK as well! 

So the latest version: @ory/kratos-client@v1.0.0 instead of v^0.13.1

If you are self-hosting you just need the kratos-client SDK. It is functionally the same as @ory/client (which is intended for [Ory Network](https://console.ory.sh/)) but if you mix them in the same project it is probably very confusing.

**Comment by foldess** [comment]
I'm having a similar problem.

In my case i did a clean installation of kratos v1.0.0 + ui from [kratos docs](https://www.ory.sh/docs/kratos/guides/deploy-kratos-example), setting cors for my app port. When testing from the UI it works fine but when using the client SDK on NEXTJS `"next": "^13.2.5-canary.28"` the .toSession() won't send any cookie thus resulting in this response:

```
{
    "error": {
        "code": 401,
        "status": "Unauthorized",
        "reason": "No valid session cookie found.",
        "message": "The request could not be authorized"
    }
}
```

I tested with `"@ory/client": "^1.1.41"` & `"@ory/kratos-client": "^1.0.0"`, but same results.

Plus i also tried setting up kratos V0.10.0 out of curiosity but got the same results.

**Reply by vinckr** [reply]
Hey @foldess 

Did you follow the [deploy guide](https://www.ory.sh/docs/kratos/guides/deploy-kratos-example) 1 to 1? 
Maybe there is some part that is outdated now, we dont have tests or automation for this quickstart.

It is more meant as a guide rail for people new to self-hosting, there are always some things you have to figure out when self hosting a service. 
That being said, can you share your code what you are doing with the SDK? 
Is the cookie being sent in the request? Can you post the full request maybe?

**Reply by foldess** [reply]
Hi @vinckr 

Yes i followed the deploy guide but with some changes: I installed is on my local machine because i'm still testing it so i'm not using any nginx configs and i'm not using any domains but localhost.

I realised while writting this, i think that using localhost with multiple ports may be the problem (because kratos is under localhost:4433, while kratos UI which sets the cookie is under localhost:4455, thus the cookies are being filtered out by same-site cookie policy)

![image](https://github.com/ory/kratos/assets/59058468/e29bea1e-1bf3-406b-8263-b00d75074816)

i'll leave my configs / codes here anyway
kratos.yml
```
version: v0.13.0

dsn: (pgstring)

serve:
  public:
    base_url: http://127.0.0.1:4433/
    cors:
      enabled: true
      allowed_origins:
        - http://127.0.0.1:3001
        - http://localhost:3001

  admin:
    base_url: http://127.0.0.1:4434/

selfservice:
  default_browser_return_url: http://127.0.0.1:4455/
  allowed_return_urls:
    - http://127.0.0.1:4455
    - http://127.0.0.1:3001
    - http://localhost:3001

  methods:
    password:
      enabled: true
    totp:
      config:
        issuer: Test
      enabled: true
    lookup_secret:
      enabled: true
    link:
      enabled: true
    code:
      enabled: true

  flows:
    error:
      ui_url: http://127.0.0.1:4455/error

    settings:
      ui_url: http://127.0.0.1:4455/settings
      privileged_session_max_age: 15m
      required_aal: highest_available

    recovery:
      enabled: true
      ui_url: http://127.0.0.1:4455/recovery
      use: code

    verification:
      enabled: true
      ui_url: http://127.0.0.1:4455/verification
      use: code
      after:
        default_browser_return_url: http://127.0.0.1:4455/

    logout:
      after:
        default_browser_return_url: http://127.0.0.1:4455/login

    login:
      ui_url: http://127.0.0.1:4455/login
      lifespan: 10m

    registration:
      lifespan: 10m
      ui_url: http://127.0.0.1:4455/registration
      after:
        password:
          hooks:
            - hook: session
            - hook: show_verification_ui

log:
  level: debug
  format: text
  leak_sensitive_values: false

secrets:
  cookie:
    - (key1)
  cipher:
    - (key2)

ciphers:
  algorithm: xchacha20-poly1305

hashers:
  algorithm: bcrypt
  bcrypt:
    cost: 8

identity:
  default_schema_id: default
  schemas:
    - id: default
      url: file:///opt/kratos/config/identity.schema.json

courier:
  smtp:
    connection_uri: smtps://test:test@mailslurper:1025/?skip_ssl_verify=true

```
code
```
import { Configuration, FrontendApi } from '@ory/kratos-client'

const ory = new FrontendApi(
	new Configuration({
		basePath: http://127.0.0.1:4433,
		baseOptions: {
			withCredentials: true,
		},
	})
)

...

useEffect(() => {
		ory.toSession()
			.then(({ data }) => {
				// User has a session!
				setSession(data)
				// Create a logout url
				ory.createBrowserLogoutFlow().then(({ data }) => {
					setLogoutUrl(data.logout_url)
				})
			})
			.catch((e) => {
				console.error(e)

				// Redirect to login page
				return router.push(`http://127.0.0.1:4455/login?return_to=${process.env.NEXT_PUBLIC_ORY_REDIRECT}`)
			})
	}, [])
```

**Reply by vinckr** [reply]
When developing locally, your application and Ory APIs must be available on the same domain (i.e., localhost) to avoid issues with third-party cookies.
If you're running your application and Ory Kratos on different ports of localhost, that should be fine as long as the domain stays the same.
Ory Kratos and your application need to be on the same domain for cookies to work properly. This is because cookies are domain-specific and cannot be shared across different domains due to the same-site cookie policy.

However, be careful not to use localhost and 127.0.0.1 interchangeably in your code. Even though they technically mean the same thing, they're treated as different hostnames. Using both can cause problems with the origin of HTTP cookies sent by the application, and Ory APIs might not be able to read the cookies sent by your app.

**Reply by foldess** [reply, marked as answer]
Thanks, i forgot that i was accessing my app at localhost while kratos was hosted at 127.0.0.1, switched to 127.0.0.1 on local and it works just fine.

edit: But i question comes to my mind: with this approach i cannot have an App running on domain A and Kratos + UI running on domain B right? would the solution be making the login from my app in domain A? Or is it a localhost only problem?

Again, thanks for your time.

**Reply by vinckr** [reply]
Glad that was it and not something else; I was running out of ideas 😁
